kimchi/src/circuits/cairo/runner/memory.rs

```diff
@@ -0,0 +1,137 @@
+//! This module represents the Cairo memory, containing the
+//! compiled Cairo program that occupies the first few entries
+
+use std::ops::{Index, IndexMut};
+
+use crate::circuits::cairo::runner::helper::*;
+use crate::circuits::cairo::runner::word::CairoWord;
+use ark_ff::PrimeField;
+use core::iter::repeat;
+use o1_utils::FieldHelpers;
+
+/// This data structure stores the memory of the program
+pub struct CairoMemory<F: PrimeField> {
+    /// length of the public memory
+    codelen: usize,
+    /// full memory vector, None if non initialized
+    pub data: Vec<Option<CairoWord<F>>>,
+}
+
+impl<F: PrimeField> Index<F> for CairoMemory<F> {
+    type Output = Option<CairoWord<F>>;
+    fn index(&self, idx: F) -> &Self::Output {
+        // Safely convert idx from F to usize (since this is a memory address
+        // idx should not be too big, this should be safe)
+        let addr: u64 = idx.to_u64();
+        &self.data[addr as usize]
+    }
+}
+
+impl<F: PrimeField> IndexMut<F> for CairoMemory<F> {
+    fn index_mut(&mut self, idx: F) -> &mut Self::Output {
+        let addr: u64 = idx.to_u64();
+        &mut self.data[addr as usize]
```

Comment:
Oh this is a very good question indeed... I hadn't think of that, since I was calling to `resize()` at the beginning of `write()` and `read()` methods. But at the end, these two are connected to the `Index` trait, so I could try to move this to `index()` and keep a clearer separation of concerns 👍